src/Microsoft.Diagnostics.DebugServices/RegisterService.cs

```diff
@@ -169,44 +171,62 @@ public unsafe bool GetRegisterValue(uint threadId, int index, out ulong value)
-                                value = *((ushort*)(ptr + info.RegisterOffset));
-                                return true;
-                            case 4:
-                                value = *((uint*)(ptr + info.RegisterOffset));
-                                return true;
-                            case 8:
-                                value = *((ulong*)(ptr + info.RegisterOffset));
-                                return true;
+                            switch (info.RegisterSize)
+                            {
+                                case 1:
+                                    value = *((byte*)(ptr + info.RegisterOffset));
+                                    return true;
+                                case 2:
+                                    value = *((ushort*)(ptr + info.RegisterOffset));
+                                    return true;
+                                case 4:
+                                    value = *((uint*)(ptr + info.RegisterOffset));
+                                    return true;
+                                case 8:
+                                    value = *((ulong*)(ptr + info.RegisterOffset));
+                                    return true;
+                            }
                         }
                     }
                 }
             }
             return false;
         }
 
-        private unsafe byte[] GetThreadContext(uint threadId)
+        /// <summary>
+        /// Returns the raw context buffer bytes for the specified thread.
+        /// </summary>
+        /// <param name="threadId">thread id</param>
+        /// <returns>register context or null if error</returns>
+        public byte[] GetThreadContext(uint threadId)
         {
             if (_threadContextCache.TryGetValue(threadId, out byte[] threadContext))
             {
                 return threadContext;
             }
             else
             {
-                threadContext = new byte[_contextSize];
-                fixed (byte* ptr = threadContext)
+                unsafe
                 {
-                    if (_target.DataReader.GetThreadContext(threadId, _contextFlags, (uint)_contextSize, new IntPtr(ptr)))
+                    threadContext = new byte[_contextSize];
+                    fixed (byte* ptr = threadContext)
                     {
-                        _threadContextCache.Add(threadId, threadContext);
-                        return threadContext;
+                        try
+                        {
+                            if (_target.DataReader.GetThreadContext(threadId, _contextFlags, (uint)_contextSize, new IntPtr(ptr)))
+                            {
+                                _threadContextCache.Add(threadId, threadContext);
+                                return threadContext;
+                            }
+                        }
+                        catch (ClrDiagnosticsException ex)
```

Comment:
Returning null on failure feels like it is going to be either annoying or error-prone for callers. Any caller that is willing to do:

```
 byte[] bla = GetThreadContext();
 if(bla == null)
 {
 handle error
}
else
{
   do work
}
````

is probably no worse off than if they wrote:

```
try
{
   byte[] bla = GetThreadContext();
   do work
}
catch(ClrDiagnosticsException ex)
{
  handle error
}
```

In the latter case they get access to a descriptive error if they want it, and if they forget to catch the exception it automatically propagates vs. getting a not so informative NullReferenceException.

Reply:
I'll revisit this interface later after 3.0 releases before any interface becomes "public".